docs/dev/commitlog-file-format.md

```diff
@@ -114,3 +114,33 @@ the file is not fully written/prematurely ended. Both cases can mean data loss,
 how writing is done as well as OS and hardware.
 
 ```
+
+Version 4
+---------
+
+Modified from v3 to allow fragmented data entries, i.e. writing a single data entry
+as a stream across several segments.
+A fragmented entry is written by splitting data into sub-parts that will fit into
+the normal restrictions of a write (i.e. smaller than max mutation size, but also
+trying to fit into existing buffers as best we can to avoid wasting alignment slack).
+
+```
+        Fragmented entry
+
+        magic           : fragmented marker - 0xfffffffe (MAX_UINT32-1)
+        size            : size of this fragmented entry part + headers
+        id              : the stream id
```

Comment:
How do you ensure its uniqueness across restarts?

Reply:
We don't. You cannot save commitlog entries across restarts. You replay and delete, or fail and retry. The id:s are unique (per shard btw) for one commitlog session.

Reply:
> You cannot save commitlog entries across restarts. You replay and delete, or fail and retry.

Doesn't hinted handoff break this assumption? It may write new segments before the old ones are processed and deleted.

Reply:
True, but as long as it does not mix the segments from different "sessions" when replaying, it is the same thing.

Reply:
We should gain the ability to checkpoint commitlog replays. This allows us to get out of OOM-like situations. But not in this series.